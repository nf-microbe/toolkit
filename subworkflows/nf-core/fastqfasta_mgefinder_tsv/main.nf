// Import modules
include { BWA_INDEX                     } from '../../../modules/nf-core/bwa/index'
include { BWA_MEM                       } from '../../../modules/nf-core/bwa/mem'
include { MGEFINDER_FORMATBAM           } from '../../../modules/nf-core/mgefinder/formatbam'
include { MGEFINDER_FIND                } from '../../../modules/nf-core/mgefinder/find'
include { MGEFINDER_PAIR                } from '../../../modules/nf-core/mgefinder/pair'
include { MGEFINDER_INFERSEQREFERENCE   } from '../../../modules/nf-core/mgefinder/inferseqreference'

workflow FASTQFASTA_MGEFINDER_TSV {

    take:
    fastq_gz            // channel: [ [ meta.id, meta.single_end ], [ path(fastq_1), path(fastq_2) ] ]
    fasta_gz            // channel: [ [ meta.id ], path(fasta)]

    main:

    ch_versions = Channel.empty()

    //
    // MODULE: Index reference genome/assembly
    //
    BWA_INDEX(
        fasta_gz
    )
    ch_versions = ch_versions.mix(BWA_INDEX.out.versions)

    // join fastQ and Fasta
    ch_bwa_mem_input = fastq_gz.map { meta, fastq -> [ [ id:meta.group ], meta, fastq ] }
        .combine(fasta_gz, by:0)
        .combine(BWA_INDEX.out.index, by:0)
        .multiMap { meta_group, meta_fastq, fastq, fasta, index ->
            fastq:  [ meta_fastq, fastq ]
            fasta:  [ meta_fastq, fasta ]
            index:  [ meta_fastq, index ]
        }

    //
    // MODULE: Align reads to reference genome/assembly
    //
    BWA_MEM(
        ch_bwa_mem_input.fastq,
        ch_bwa_mem_input.index,
        ch_bwa_mem_input.fasta,
        false
    )
    ch_versions = ch_versions.mix(BWA_MEM.out.versions)

    //
    // MODULE: Format BAM
    //
    MGEFINDER_FORMATBAM(
        BWA_MEM.out.bam
    )
    ch_versions = ch_versions.mix(MGEFINDER_FORMATBAM.out.versions)

    //
    // MODULE: Find insertion sites
    //
    MGEFINDER_FIND(
        MGEFINDER_FORMATBAM.out.bam
    )
    ch_versions = ch_versions.mix(MGEFINDER_FIND.out.versions)

    ch_mgefinder_find = MGEFINDER_FIND.out.find
        .filter { meta, find ->
            find.countLines( limit: 10 ) > 1
        }

    // join fasta, find bam, and find file
    ch_pair_input = fasta_gz
        .combine(MGEFINDER_FORMATBAM.out.bam.map { meta, bam -> [ [ id: meta.group ], meta, bam ] }, by:0)
        .combine(ch_mgefinder_find.map { meta, find -> [ [ id: meta.group ], find ] }, by:0)
        .multiMap { meta_group, fasta, meta_bam, bam, find ->
            fasta:  [ meta_bam, fasta ]
            bam:    [ meta_bam, bam ]
            find:   [ meta_bam, find ]
        }

    //
    // MODULE: Pair termini with each other
    //
    MGEFINDER_PAIR(
        ch_pair_input.fasta,
        ch_pair_input.bam,
        ch_pair_input.find
    )
    ch_versions = ch_versions.mix(MGEFINDER_PAIR.out.versions.first())

    ch_mgefinder_pair = MGEFINDER_PAIR.out.pair
        .filter { meta, pair ->
            pair.countLines( limit: 10 ) > 1
        }

    // join FastA and pair file
    ch_inferseqreference_input = fasta_gz
        .combine(ch_mgefinder_pair.map { meta, pair -> [ [ id: meta.group ], meta, pair ] }, by:0)
        .multiMap { meta_group, fasta, meta_pair, pair ->
            fasta:  [ meta_pair, fasta ]
            pair:   [ meta_pair, pair ]
        }

    //
    // MODULE: infer the identity of insertions
    //
    MGEFINDER_INFERSEQREFERENCE(
        ch_inferseqreference_input.fasta,
        ch_inferseqreference_input.pair
    )

    emit:
    prophage    = MGEFINDER_INFERSEQREFERENCE.out.tsv   // channel: [ [ meta ], inferse_reference.tsv ]
    versions    = ch_versions                           // channel: [ versions.yml ]
}
