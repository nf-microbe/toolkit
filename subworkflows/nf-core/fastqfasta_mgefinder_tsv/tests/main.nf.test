nextflow_workflow {

    name "Test Subworkflow FASTQFASTA_MGEFINDER_TSV"
    script "../main.nf"
    workflow "FASTQFASTA_MGEFINDER_TSV"
    config "./nextflow.config"

    tag "subworkflows"
    tag "subworkflows_nfcore"
    tag "subworkflows/fastqfastamgefinder_tsv"
    tag "bwa/index"
    tag "bwa/mem"
    tag "mgefinder/formatbam"
    tag "mgefinder/find"
    tag "mgefinder/pair"
    tag "mgefinder/inferseqreference"

    test("b_fragilis - fastq.gz - fasta.gz") {
        when {
            workflow {
                """
                input[0] = Channel.of([
                    [ id:'test', group:'test', single_end:false ],
                    [
                        file("/mmfs1/gscratch/pedslabs_hoffman/carsonjm/projects/atb_prophage_eval/results/sra_asperacli/SAMEA5584860_1.fastq.gz", checkIfExists: true),
                        file("/mmfs1/gscratch/pedslabs_hoffman/carsonjm/projects/atb_prophage_eval/results/sra_asperacli/SAMEA5584860_2.fastq.gz", checkIfExists: true)
                    ]
                ])
                input[1] = Channel.of([
                    [ id:'test' ],
                    file("/mmfs1/gscratch/pedslabs_hoffman/carsonjm/projects/atb_prophage_eval/results/allthebacteria_aria2seqkit/SAMEA5584860.fasta.gz", checkIfExists: true)
                ])
                """
            }
        }
        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
