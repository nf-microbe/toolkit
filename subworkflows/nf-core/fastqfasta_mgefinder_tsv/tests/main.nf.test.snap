{
    "b_fragilis - fastq.gz - fasta.gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.prophage.out:md5,8166ad81eb844e4172fd02fb2ec0db2f"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.prophage.out:md5,8166ad81eb844e4172fd02fb2ec0db2f"
                    ]
                ],
                "1": [
                    "versions.yml:md5,2918455fb750db2efde07a9ea12b6ceb",
                    "versions.yml:md5,45815fe0f7dbec34978e0ad426ac94cb",
                    "versions.yml:md5,45815fe0f7dbec34978e0ad426ac94cb",
                    "versions.yml:md5,714c151644dc1370106db50ec068ac08",
                    "versions.yml:md5,e21076525aaa5f7e9005eb2bee4e4668",
                    "versions.yml:md5,e21076525aaa5f7e9005eb2bee4e4668"
                ],
                "prophage": [
                    [
                        {
                            "id": "test1",
                            "group": "test",
                            "single_end": false
                        },
                        "test1.prophage.out:md5,8166ad81eb844e4172fd02fb2ec0db2f"
                    ],
                    [
                        {
                            "id": "test2",
                            "group": "test",
                            "single_end": false
                        },
                        "test2.prophage.out:md5,8166ad81eb844e4172fd02fb2ec0db2f"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,2918455fb750db2efde07a9ea12b6ceb",
                    "versions.yml:md5,45815fe0f7dbec34978e0ad426ac94cb",
                    "versions.yml:md5,45815fe0f7dbec34978e0ad426ac94cb",
                    "versions.yml:md5,714c151644dc1370106db50ec068ac08",
                    "versions.yml:md5,e21076525aaa5f7e9005eb2bee4e4668",
                    "versions.yml:md5,e21076525aaa5f7e9005eb2bee4e4668"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-21T08:05:54.547327855"
    }
}